opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/batcher.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Sequence, Type
+
+from opentelemetry.metrics import Counter, MetricT
+from opentelemetry.sdk.metrics.export import MetricRecord
+from opentelemetry.sdk.metrics.export.aggregate import (
+    Aggregator,
+    CounterAggregator,
+)
+
+
+class Batcher(abc.ABC):
+    """Base class for all batcher types.
+
+    The batcher is responsible for storing the aggregators and aggregated
+    values received from updates from metrics in the meter. The stored values will be
+    sent to an exporter for exporting.
+    """
+
+    def __init__(self, keep_state: bool):
```

Comment:
Are you referring to the fact that the constructor has the `keep_state` parameter or the fact that the `Batcher` abc has real methods (`aggregator_for`, `check_point_set`, etc)?

Reply:
Both! If you move the implementations for those methods to another class I assume you'd move the constructor too.

Reply:
How would I be able to have implementations for classes that share the same behaviour? For example, if `aggregator_for` works the same for all types of `Batcher`, I should be able to define the implementation within the parent class no?

Reply:
In general you could do this with a separate base class (e.g. `BaseBatcher`) that extends the ABC class or a with mixin (e.g. `AggregatorSelector`) that adds the particular methods you need. But that assumes you think it's worth the trouble to have separate interface and implementation classes.

A lot of the discussion about interfaces in the API in #66 and #311 is relevant here too. In both the API and SDK packages, each ABC class represents a component that an alternative SDK could replace. That is, the `foovendor-opentelemetry-sdk` package could contain a single custom `FooVendorBatcher` class that extends `Batcher`. Concrete methods in ABCs make writing replacement classes more difficult since the author needs to understand the implementation to know which methods to override. And as in https://github.com/open-telemetry/opentelemetry-python/issues/66#issuecomment-565276593, this can lead to some surprising results (instead of an outright crash) if there are changes to the API.

One interesting exception here is convenience methods like [`Context.use` from #395](https://github.com/codeboten/opentelemetry-python/blob/15805a4b4103199c31902cd271e9a5869e222a27/opentelemetry-api/src/opentelemetry/context/context.py#L58-L65) (since removed, but still a great example!) that call other abstract methods, but don't assume anything else about the implementation.

But I think there's another problem here with `aggregator_for`: if it works the same way for every batcher type, why is it part of this class in the first place?